// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: timekeeping_service/api/data.proto

package api

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
)

// Validate checks the field values on GetHistoryOfUserRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *GetHistoryOfUserRequest) Validate() error {
	if m == nil {
		return nil
	}

	if m.GetUserId() <= 0 {
		return GetHistoryOfUserRequestValidationError{
			field:  "UserId",
			reason: "value must be greater than 0",
		}
	}

	if v, ok := interface{}(m.GetTime()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetHistoryOfUserRequestValidationError{
				field:  "Time",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// GetHistoryOfUserRequestValidationError is the validation error returned by
// GetHistoryOfUserRequest.Validate if the designated constraints aren't met.
type GetHistoryOfUserRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetHistoryOfUserRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetHistoryOfUserRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetHistoryOfUserRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetHistoryOfUserRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetHistoryOfUserRequestValidationError) ErrorName() string {
	return "GetHistoryOfUserRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetHistoryOfUserRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetHistoryOfUserRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetHistoryOfUserRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetHistoryOfUserRequestValidationError{}

// Validate checks the field values on GetHistoryOfUserResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *GetHistoryOfUserResponse) Validate() error {
	if m == nil {
		return nil
	}

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return GetHistoryOfUserResponseValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	return nil
}

// GetHistoryOfUserResponseValidationError is the validation error returned by
// GetHistoryOfUserResponse.Validate if the designated constraints aren't met.
type GetHistoryOfUserResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetHistoryOfUserResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetHistoryOfUserResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetHistoryOfUserResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetHistoryOfUserResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetHistoryOfUserResponseValidationError) ErrorName() string {
	return "GetHistoryOfUserResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetHistoryOfUserResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetHistoryOfUserResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetHistoryOfUserResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetHistoryOfUserResponseValidationError{}

// Validate checks the field values on HistoryTime with the rules defined in
// the proto definition for this message. If any rules are violated, an error
// is returned.
func (m *HistoryTime) Validate() error {
	if m == nil {
		return nil
	}

	if m.GetFrom() <= 0 {
		return HistoryTimeValidationError{
			field:  "From",
			reason: "value must be greater than 0",
		}
	}

	if m.GetTo() <= 0 {
		return HistoryTimeValidationError{
			field:  "To",
			reason: "value must be greater than 0",
		}
	}

	return nil
}

// HistoryTimeValidationError is the validation error returned by
// HistoryTime.Validate if the designated constraints aren't met.
type HistoryTimeValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e HistoryTimeValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e HistoryTimeValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e HistoryTimeValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e HistoryTimeValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e HistoryTimeValidationError) ErrorName() string { return "HistoryTimeValidationError" }

// Error satisfies the builtin error interface
func (e HistoryTimeValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sHistoryTime.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = HistoryTimeValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = HistoryTimeValidationError{}

// Validate checks the field values on TimekeepingHistory with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *TimekeepingHistory) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for UserId

	if val := m.GetDay(); val <= 0 || val >= 32 {
		return TimekeepingHistoryValidationError{
			field:  "Day",
			reason: "value must be inside range (0, 32)",
		}
	}

	if val := m.GetMonth(); val <= 0 || val >= 13 {
		return TimekeepingHistoryValidationError{
			field:  "Month",
			reason: "value must be inside range (0, 13)",
		}
	}

	if m.GetYear() <= 0 {
		return TimekeepingHistoryValidationError{
			field:  "Year",
			reason: "value must be greater than 0",
		}
	}

	if v, ok := interface{}(m.GetData()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TimekeepingHistoryValidationError{
				field:  "Data",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// TimekeepingHistoryValidationError is the validation error returned by
// TimekeepingHistory.Validate if the designated constraints aren't met.
type TimekeepingHistoryValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TimekeepingHistoryValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TimekeepingHistoryValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TimekeepingHistoryValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TimekeepingHistoryValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TimekeepingHistoryValidationError) ErrorName() string {
	return "TimekeepingHistoryValidationError"
}

// Error satisfies the builtin error interface
func (e TimekeepingHistoryValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTimekeepingHistory.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TimekeepingHistoryValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TimekeepingHistoryValidationError{}

// Validate checks the field values on Data with the rules defined in the proto
// definition for this message. If any rules are violated, an error is returned.
func (m *Data) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for TotalTime

	// no validation rules for StartTime

	// no validation rules for EndTime

	for idx, item := range m.GetDetails() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataValidationError{
					field:  fmt.Sprintf("Details[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	return nil
}

// DataValidationError is the validation error returned by Data.Validate if the
// designated constraints aren't met.
type DataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataValidationError) ErrorName() string { return "DataValidationError" }

// Error satisfies the builtin error interface
func (e DataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sData.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataValidationError{}

// Validate checks the field values on CreateHistoryOfUserRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *CreateHistoryOfUserRequest) Validate() error {
	if m == nil {
		return nil
	}

	if m.GetUserId() < 0 {
		return CreateHistoryOfUserRequestValidationError{
			field:  "UserId",
			reason: "value must be greater than or equal to 0",
		}
	}

	if m.GetStartTime() <= 0 {
		return CreateHistoryOfUserRequestValidationError{
			field:  "StartTime",
			reason: "value must be greater than 0",
		}
	}

	if m.GetEndTime() <= 0 {
		return CreateHistoryOfUserRequestValidationError{
			field:  "EndTime",
			reason: "value must be greater than 0",
		}
	}

	return nil
}

// CreateHistoryOfUserRequestValidationError is the validation error returned
// by CreateHistoryOfUserRequest.Validate if the designated constraints aren't met.
type CreateHistoryOfUserRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateHistoryOfUserRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateHistoryOfUserRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateHistoryOfUserRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateHistoryOfUserRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateHistoryOfUserRequestValidationError) ErrorName() string {
	return "CreateHistoryOfUserRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateHistoryOfUserRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateHistoryOfUserRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateHistoryOfUserRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateHistoryOfUserRequestValidationError{}

// Validate checks the field values on CreateHistoryOfUserResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *CreateHistoryOfUserResponse) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Code

	// no validation rules for Message

	return nil
}

// CreateHistoryOfUserResponseValidationError is the validation error returned
// by CreateHistoryOfUserResponse.Validate if the designated constraints
// aren't met.
type CreateHistoryOfUserResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateHistoryOfUserResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateHistoryOfUserResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateHistoryOfUserResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateHistoryOfUserResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateHistoryOfUserResponseValidationError) ErrorName() string {
	return "CreateHistoryOfUserResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateHistoryOfUserResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateHistoryOfUserResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateHistoryOfUserResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateHistoryOfUserResponseValidationError{}

// Validate checks the field values on Data_Line with the rules defined in the
// proto definition for this message. If any rules are violated, an error is returned.
func (m *Data_Line) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Index

	// no validation rules for StartTime

	// no validation rules for EndTime

	return nil
}

// Data_LineValidationError is the validation error returned by
// Data_Line.Validate if the designated constraints aren't met.
type Data_LineValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e Data_LineValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e Data_LineValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e Data_LineValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e Data_LineValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e Data_LineValidationError) ErrorName() string { return "Data_LineValidationError" }

// Error satisfies the builtin error interface
func (e Data_LineValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sData_Line.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = Data_LineValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = Data_LineValidationError{}
